### ISSUE TITLE
Unexpected results with empty title element

### ISSUE THREAD

**Issue by daurnimator** [craigbarnes/lua-gumbo#42]
It appears that lua-gumbo will emit an empty document if it encounters an empty title element:

```
$ lua5.1  -e 'print(require"gumbo".parse("<head><title/></head><body>hi</body>").body.outerHTML)'
<body></body>
```

```
$ lua5.1  -e 'print(require"gumbo".parse("<head><title>asd</title></head><body>hi</body>").body.outerHTML)'
<body>hi</body>
```

Not sure if this is an issue with gumbo in general or just lua-gumbo

**Comment by craigbarnes** [comment]
The `<title/>` in your first example puts the parser into a (standardized) error state and causes it consume the rest of the input. You can see this by serializing the `documentElement` node (the `<html>` Element; inserted automatically in this case) instead of the `body` node:

```lua
local gumbo = require "gumbo"
local document = assert(gumbo.parse("<head><title/></head><body>hi</body>"))
print(document.documentElement.outerHTML)
```

which, for me, produces:

```html
<html><head><title>&lt;/head&gt;&lt;body&gt;hi&lt;/body&gt;</title></head><body></body></html>
```

Firefox 37 and Chrome 40 both give the same output for `document.documentElement.outerHTML`.

Basically `<title/>` is tolerated by the parser but it's not valid syntax. HTML5 doesn't have "self-closing" tags like XML does.

**Comment by daurnimator** [comment]
Ah. thanks for the explanation.

Is there some mode where the xhtml style is accepted? This doesn't seem to work (i'm not sure if it's meant to).

```
$ lua5.1  -e 'print(require"gumbo".parse([[<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html><head><title/></head><body>hi</body></html>]]).documentElement.outerHTML)'
<html><head><title>&lt;/head&gt;&lt;body&gt;hi&lt;/body&gt;&lt;/html&gt;</title></head><body></body></html>
```

**Comment by craigbarnes** [comment]
> Is there some mode where the xhtml style is accepted?

Not really. The [HTML vs XHTML](https://html.spec.whatwg.org/multipage/introduction.html#html-vs-xhtml) page of the spec explains the differences quite well. Basically, it's the MIME/content type rather than the doctype that determines which parser should be used. Only documents served as `application/xhtml+xml` will be recognized as XHTML and in that case the spec calls for a real XML parser to be used.

I'm open to suggestions on how this could be made more straight forward for a library user. Moving the DOM library to a separate module (#39) would open the door for lua-gumbo and some yet-to-be-written XML bindings to share the same tree format and satisfy both parts of the spec.

**Comment by daurnimator** [comment]
So I guess I'll close this as not a bug?
